### ISSUE TITLE
Invalid Cast Exception when running code interactive on Mono

### ISSUE THREAD

**Issue by boeremak** [fsharp/fsharp#423]
Hi I don't know if this is the right place, but  I am going to post it here because I don't know where else to report it.
When I run the following code in FSharpi on Mono

```
open System
open System.Diagnostics


type Option<'a> =
    | None
    | Some of 'a

type Result<'TSuccess,'TFailure>=
    | Success of 'TSuccess
    | Failure of 'TFailure

let Trimchars = if Environment.OSVersion.Platform.Equals("Unix") then "\n" else "\r\n"


let RunCommand command parameters =    
    let startInfo = new ProcessStartInfo()
    startInfo.CreateNoWindow <- true;
    startInfo.UseShellExecute <-false;
    startInfo.FileName <- command
    startInfo.Arguments <- parameters
    startInfo.RedirectStandardOutput <- true;
    use proc = Process.Start(startInfo);
    let output = proc.StandardOutput.ReadToEnd();
    proc.WaitForExit()
    output.TrimEnd(Trimchars.ToCharArray())

let GetZpoolNames() = 
    let output = RunCommand "/sbin/zpool" "list"
                    |> (fun x -> x.Split(Trimchars.ToCharArray(),StringSplitOptions.RemoveEmptyEntries))
                    |> List.ofArray
    
    if  output.Head |> (fun x -> x.Equals("no pools available"))
    then None
    else output
         |> List.tail
         |> List.map(fun x -> x.Substring(0, x.IndexOf(" ")))
         |> (fun x -> Some(x))

```

When I run:
```
GetZpoolNames()
```

I get the following error:

```
Please report this so it can be fixed.ing list>
Trace: System.InvalidCastException: Cannot cast from source type to destination type.
  at (wrapper castclass) object:__castclass_with_cache (object,intptr,intptr)
  at Microsoft.FSharp.Compiler.Interactive.Shell+Utilities+AnyToLayoutSpecialization`1[FSI_0002+Option`1[Microsoft.FSharp.Collections.FSharpList`1[System.String]]].Microsoft-FSharp-Compiler-Interactive-Shell-Utilities-IAnyToLayoutCall-AnyToLayout (Internal.Utilities.StructuredFormat.FormatOptions options, System.Object o) [0x00000] in <filename unknown>:0
  at Microsoft.FSharp.Compiler.Interactive.Shell+FsiValuePrinter.PrintValue (Microsoft.FSharp.Compiler.Interactive.FsiValuePrinterMode printMode, Internal.Utilities.StructuredFormat.FormatOptions opts, System.Object x, System.Type ty) [0x00000] in <filename unknown>:0                                              val it : Option<string list> =

```
If I compile the program it runs fine.

```
output:
pool
storagepool
```

Am I doing something wrong? 

Info:
``` 
Debian 8.1 Linux
Mono:

Mono JIT compiler version 4.0.1 (tarball Thu May 28 09:08:28 UTC 2015)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        LLVM:          supported, not enabled.
        GC:            sgen
```

**Comment by dsyme** [comment]
Looks related to #426.  Could you add a Mono/Xamarin buzilla bug on this please?

**Comment by obadz** [comment]
Very likely fixed by https://github.com/mono/mono/pull/1654 but hard to verify without /sbin/zpool

**Comment by boeremak** [comment]
I think I figured it out, after trying to recreate it with some standard output pasted as a variable it worked.
The self defined option type was tripping it up.

type Option<'a> =
    | None
    | Some of 'a

This trips up the runtime.

My apologies if this is normal behaviour.

**Comment by dsyme** [comment]
@boeremak - do you have a standalone repro for how the self-defined option type causes problems for the runtime?  It's possible that using a self-defined option type exposed you to the bug fixed by mono/mono#1654 - but with that fix in place it there should be no problem with using a self-defined option type.

**Comment by dsyme** [comment]
(closing this as we believe it fixed by mono/mono#1654)
